File: mint-runtime-permission/src/main/java/net/crystalapps/permission/runtime/library/core/PermProxyInvocationHandler.java
package net.crystalapps.permission.runtime.library.core;

import android.support.annotation.NonNull;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;

/**
 * Created by Syed Owais Ali on 10/13/2018.
 */
public class PermProxyInvocationHandler<T> implements InvocationHandler {

    @NonNull
    private final String type;
    private final int requestCode;
    @NonNull
    private final T caller;

    public PermProxyInvocationHandler(@NonNull String type, int requestCode, @NonNull T caller) {
        this.type = type;
        this.requestCode = requestCode;
        this.caller = caller;
    }

    @Override
    public Object invoke(Object o, Method method, Object[] objects) {

        switch (method.getName()) {
            case "getType":
                return type;
            case "getRequestCode":
                return requestCode;
            case "getCaller":
                return caller;

            default:
                throw new UnsupportedOperationException("Unsupported method: " + method.getName());
        }
    }
}

File: mint-runtime-permission/src/main/java/net/crystalapps/permission/runtime/library/core/PermProxyInvocationHandler.kt
package net.crystalapps.permission.runtime.library.core

import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method

/**
 * Created by Syed Owais Ali on 10/13/2018.
 */
class PermProxyInvocationHandler<T>(private val type: String, private val requestCode: Int, private val caller: T) : InvocationHandler {

    override fun invoke(o: Any?, method: Method?, objects: Array<out Any>?): Any {

        return when (method?.name) {
            "getType" -> type
            "getRequestCode" -> requestCode
            "getCaller" -> caller as Any

            else -> throw UnsupportedOperationException("Unsupported method: " + method?.name)
        }
    }
}
